keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+      expression: "spec.list.map(x, x.metadata)"
+    - name: itemMetadataNames
+      expression: "metadataList.map(m, m.name)"
+  validations:
+    - expression: "itemMetadataNames.all(name, name.startsWith('xyz-'))"
+    - expression: "itemMetadataNames.exists(name, name == 'required')"
+```
+
+##### Scopes
+
+<<[UNRESOLVED jpbetz, ?? ]>> 
+This is not yet agreed on, and some reviewers have found it more
+difficult to understand validations using this feature than validations that do
+not.
+
+Also, the "bind varables" idea is maybe too ambitious?
+<<[/UNRESOLVED]>>
+
+Imagine that a policy validation uses a CEL expression find an invalid value in
+a list somewhere nested in a resource.  E.g. `spec.initContainers.<listitem>.name`.
+
+How do we:
+
+- Including the invalid value in a message?
+  - Using a second expression to build a message that then must again find the
+    validation error in the list duplicates a lot of code, and is inefficient.
+- Include the field path of the problem field in the message?
+  - It can be hard coded in the message string for basic cases, but for more
+    complex cases (where the are map keys or array indices involved) it becomes
+    messy and complicated to reconstruct.
+- Traverse across fields safely? CEL offers the `all()` macro, so traversals
+  like `spec.initContainers.all(c, c.name)` are possible, but can be subtly
+  incorrect because of optional fields (`initContainers` in this case), which
+  require `has()` checks, e.g.: `!has(spec.initContainers) || spec.initContainers.all(c, c.name)`.
+
+CRD Validation Rules solved these problems by allowing validation rules to
+attached to any schema in the OpenAPIv3. The validation rules are scoped to
+whatever location in the OpenAPIv3 they are attached.
+
+We propose using a simple schema path format. The purpose of the path is to
+uniquely identify a schema from the root of a CRDs OpenAPIv3 schema.
+
+Example:
+
+```
+spec.initContainers         # Schema of initContainers array
+spec.initContainers[*]      # Schema of the items of the initContainers array
+spec.initContainers[*].name # Schema of the name of initContainers
+```
+
+For example, to validate all containers:
+
+```yaml
+  validations:
+    - scope: "spec.containers[*]"
+      expression: "scope.name.startsWith('xyz-')"
+      message: "{1} does not start with 'xyz'"
+      messageArgs: ["scope.name"]
+```
+
+To make it possible to access the path information in the scope, we can offer a
+way to bind varables to the map and list indices in the path, e.g.:
+
+```
+spec.x[xKey].y[yIndex].field
+```
+
+```yaml
+  validations:
+    - scope: "x[xKey].y[yIndex].field"
+      expression: "scope.startsWith('xyz-')"
+      message: "{1}, {2}: some problem"
+      messageArgs: ["scopePath.xKey", "scopePath.yIndex"]
+```
+
+Prior art:
+
+- cel-policy-template's offer a
+  [`range`](https://github.com/google/cel-policy-templates-go/blob/master/test/testdata/map_ranges/template.yaml)
```

Comment:
FWIW `range` actually binds variables to the value of a field at the range, the CEL expression isn't scoped.

This is significantly more powerful, as it allows you to perform checks like "only containers named 'server' have a port named 'web' that listens on port '80'"

To do the same with scope expressions,  you'd need fancier syntax (e.g. `spec.initContainers[!~^server$]` to check a regex or similar), probably something best left to CEL.

Reply:
I like `range`, but think it's good as a post-alpha feature only because it's possible that some of what's offered in `range` may become a part of core CEL. 

One deficit of CEL is that macros only give you access to the list element value, but not the index (reverse for maps), and if the key, value pair were available in both contexts, then you can express what `range` does inside CEL. Currently, you can't.

Reply:
I think variables are good enough in the short term. I'll make note of range in the future work section.